src/backend/distributed/operations/citus_tools.c

```diff
@@ -54,12 +57,40 @@ static void ExecuteCommandsAndStoreResults(StringInfo *nodeNameArray,
 										   int commandCount);
 static bool ExecuteRemoteQueryOrCommand(char *nodeName, uint32 nodePort,
 										char *queryString, StringInfo queryResult);
+static bool ExecuteRemoteQueryOrCommandUsingExistingConnections(char *nodeName,
+																uint32 nodePort,
+																char *queryString,
+																StringInfo
+																queryResultString);
+static bool ExecuteRemoteQueryOrCommandInternal(char *nodeName, uint32 nodePort,
+												char *queryString,
+												StringInfo queryResultString,
+												int connectionFlags);
 static Tuplestorestate * CreateTupleStore(TupleDesc tupleDescriptor,
 										  StringInfo *nodeNameArray, int *nodePortArray,
 										  bool *statusArray,
 										  StringInfo *resultArray, int commandCount);
 
 
+/*
+ * worker_check_connection_to_node sends a simple query from a worker node to another
+ * node, and returns success status.
+ */
+Datum
+worker_check_connection_to_node(PG_FUNCTION_ARGS)
+{
+	char *nodeName = PG_GETARG_TEXT_TO_CSTRING(0);
+	int32 nodePort = PG_GETARG_INT32(1);
+
+	StringInfo queryResultString = makeStringInfo();
+	bool success =
+		ExecuteRemoteQueryOrCommandUsingExistingConnections(nodeName, nodePort,
+															CONNECTIVITY_CHECK_QUERY,
```

Comment:
ExecuteRemoteQueryOrCommandInternal uses `raiseInterrupts` as true, meaning that it might throw an error. I think, in our case, this UDF should not throw any errors but return `false`.  So, we might use `PG_TRY/PG_CATCH` to be confident that we never 

Also, we should have failure tests for this UDF, potentially on every stage: connection establishment, before query sent, after query sent etc. I forgot all the stages, but you can check the failure test source files to find about them

Reply:
I gathered the network traffic during 

```sql
SELECT citus.mitmproxy('conn.allow()');
SELECT citus.clear_network_traffic();
SELECT * FROM check_connection_to_node('localhost', :worker_2_proxy_port);
SELECT * FROM citus.dump_network_traffic();
```
result:
```
SELECT * FROM citus.dump_network_traffic();
 conn |   source    |                                                                                                                                                                                                                                                                                                                      message
---------------------------------------------------------------------
    8 | worker      | ['AuthenticationOk()', 'ParameterStatus(application_name=citus)', 'ParameterStatus(client_encoding=UTF8)', 'ParameterStatus(DateStyle=ISO, MDY)', 'ParameterStatus(default_transaction_read_only=off)', 'ParameterStatus(in_hot_standby=off)', 'ParameterStatus(integer_datetimes=on)', 'ParameterStatus(IntervalStyle=postgres)', 'ParameterStatus(is_superuser=on)', 'ParameterStatus(server_encoding=UTF8)', 'ParameterStatus(server_version=XXX)', 'ParameterStatus(session_authorization=postgres)', 'ParameterStatus(standard_conforming_strings=on)', 'ParameterStatus(TimeZone=XXX)', 'BackendKeyData(XXX)', 'ReadyForQuery(state=idle)']
    0 | coordinator | [initial message]
    1 | coordinator | [initial message]
    2 | coordinator | [initial message]
    1 | worker      | ['AuthenticationOk()', 'ParameterStatus(application_name=citus)', 'ParameterStatus(client_encoding=UTF8)', 'ParameterStatus(DateStyle=ISO, MDY)', 'ParameterStatus(default_transaction_read_only=off)', 'ParameterStatus(in_hot_standby=off)', 'ParameterStatus(integer_datetimes=on)', 'ParameterStatus(IntervalStyle=postgres)', 'ParameterStatus(is_superuser=on)', 'ParameterStatus(server_encoding=UTF8)', 'ParameterStatus(server_version=XXX)', 'ParameterStatus(session_authorization=postgres)', 'ParameterStatus(standard_conforming_strings=on)', 'ParameterStatus(TimeZone=XXX)', 'BackendKeyData(XXX)', 'ReadyForQuery(state=idle)']
    0 | worker      | ['AuthenticationOk()', 'ParameterStatus(application_name=citus)', 'ParameterStatus(client_encoding=UTF8)', 'ParameterStatus(DateStyle=ISO, MDY)', 'ParameterStatus(default_transaction_read_only=off)', 'ParameterStatus(in_hot_standby=off)', 'ParameterStatus(integer_datetimes=on)', 'ParameterStatus(IntervalStyle=postgres)', 'ParameterStatus(is_superuser=on)', 'ParameterStatus(server_encoding=UTF8)', 'ParameterStatus(server_version=XXX)', 'ParameterStatus(session_authorization=postgres)', 'ParameterStatus(standard_conforming_strings=on)', 'ParameterStatus(TimeZone=XXX)', 'BackendKeyData(XXX)', 'ReadyForQuery(state=idle)']
    2 | worker      | ['AuthenticationOk()', 'ParameterStatus(application_name=citus)', 'ParameterStatus(client_encoding=UTF8)', 'ParameterStatus(DateStyle=ISO, MDY)', 'ParameterStatus(default_transaction_read_only=off)', 'ParameterStatus(in_hot_standby=off)', 'ParameterStatus(integer_datetimes=on)', 'ParameterStatus(IntervalStyle=postgres)', 'ParameterStatus(is_superuser=on)', 'ParameterStatus(server_encoding=UTF8)', 'ParameterStatus(server_version=XXX)', 'ParameterStatus(session_authorization=postgres)', 'ParameterStatus(standard_conforming_strings=on)', 'ParameterStatus(TimeZone=XXX)', 'BackendKeyData(XXX)', 'ReadyForQuery(state=idle)']
    2 | coordinator | ['Query(query=SELECT 1)']
    2 | worker      | ["RowDescription(fieldcount=1,fields=['F(name=?column?,tableoid=0,colattrnum=0,typoid=23,typlen=4,typmod=-1,format_code=0)'])", 'DataRow(columncount=1,columns=["C(length=0,value=b\'\')"])', 'CommandComplete(command=SELECT 1)', 'ReadyForQuery(state=idle)']
    3 | coordinator | [initial message]
(10 rows)
```

If I filter out all the data excepts for those on connection 2
```
 conn |   source    |                                                                                                                                                                                                                                                                                                                      message
---------------------------------------------------------------------
    2 | coordinator | [initial message]
    2 | worker      | ['AuthenticationOk()', 'ParameterStatus(application_name=citus)', 'ParameterStatus(client_encoding=UTF8)', 'ParameterStatus(DateStyle=ISO, MDY)', 'ParameterStatus(default_transaction_read_only=off)', 'ParameterStatus(in_hot_standby=off)', 'ParameterStatus(integer_datetimes=on)', 'ParameterStatus(IntervalStyle=postgres)', 'ParameterStatus(is_superuser=on)', 'ParameterStatus(server_encoding=UTF8)', 'ParameterStatus(server_version=XXX)', 'ParameterStatus(session_authorization=postgres)', 'ParameterStatus(standard_conforming_strings=on)', 'ParameterStatus(TimeZone=XXX)', 'BackendKeyData(XXX)', 'ReadyForQuery(state=idle)']
    2 | coordinator | ['Query(query=SELECT 1)']
    2 | worker      | ["RowDescription(fieldcount=1,fields=['F(name=?column?,tableoid=0,colattrnum=0,typoid=23,typlen=4,typmod=-1,format_code=0)'])", 'DataRow(columncount=1,columns=["C(length=0,value=b\'\')"])', 'CommandComplete(command=SELECT 1)', 'ReadyForQuery(state=idle)']
```

I added tests only for `Query` and `CommandComplete` messages. Do you think that is enough @onderkalaci ?

Reply:
It'd be nice to have `AuthenticationOk ` check as well if easy to add?